Add toc_pages plugin

okkez opened the pull request:
This is a daimon_markdown plugin.

okkez committed:
Add toc_pages plugin

This is a daimon_markdown plugin.

okkez committed:
Show toc in first page only

okkez committed:
Set id to header nodes properly

okkez committed:
Search header node immediately after comment node strictly

okkez committed:
Add test for toc_pages daimon_markdown plugin

mtsmfm commented on lib/daimon_markdown/plugins/toc_pages.rb:
```diff
@@ -0,0 +1,101 @@
+
+        request = context[:request]
+
+        current_page = if request.params.key?(:page)
```
How about passing context expressly (`:page`, `:public_id` etc) instead of `:request`?
It is difficult for me to find what params this plugin needs 😕

okkez replied:
I want to use full functionalities of `request` in this plugin.
But in fact, this plugin uses some functionalities of `request`.

mtsmfm reviewed (commented):
Great work!
But `TableOfContentsPages#call` method is difficult and complex for me 😢

okkez committed:
Stop using request object in toc_pages

okkez committed:
Set context when preview

okkez commented:
I've fixed and pushed changes.
@mtsmfm Could you check this PR again?

mtsmfm commented on app/controllers/editor/posts_controller.rb:
```diff
@@ -57,6 +57,12 @@ def preview
 
+    @context = {
+      full_text: @post.body,
+      fullpath: request.fullpath,
```
Hmm, I want to have uniform name
(`fulltext` and `fullpath`) or (`full_text` and `full_path`)
but `fulltext`doesn't make sence...

mtsmfm commented on app/controllers/editor/posts_controller.rb:
```diff
@@ -57,6 +57,12 @@ def preview
         @post.pages.page(params[:page]).per(1)
       end
 
+    @context = {
```
How about `markdown_context` ?
`markdown_context` is straightforward I think.

okkez replied:
Fixed.

okkez committed:
Rename variable

@context ->
@markdown_context

It's straightforward.

okkez committed:
Rename context key name `full_text` to `original_text`

Because `full_text` and `fullpath` don't have same form.
It's very confusing us.

mtsmfm committed:
Merge branch 'master' into toc-pages

okkez committed:
Disable rubocop metrics inline

okkez committed:
Use String#tr! instead of String#gsub!

okkez committed:
Fix conditional assignment

okkez committed:
Use double quote instead of %Q notation

okkez committed:
Disable Style/NumericPredicate

okkez committed:
Disable Metrics/BlockLength in test

okkez committed:
Fix String#tr! usage

mtsmfm commented on lib/daimon_markdown/plugins/toc_pages.rb:
```diff
@@ -0,0 +1,107 @@
+      # rubocop:disable Metrics/MethodLength
+      # rubocop:disable Metrics/CyclomaticComplexity
+      # rubocop:disable Metrics/AbcSize
+      # rubocop:disable Metrics/PerceivedComplexity
```
:cry:

mtsmfm commented on lib/daimon_markdown/plugins/toc_pages.rb:
```diff
@@ -0,0 +1,99 @@
+          if header_content
+            diff = level - previous_level
+            case
+            when diff > 0
```
:robot:

okkez replied:
Disable in .rubocop.yml

okkez commented on lib/daimon_markdown/plugins/toc_pages.rb:
```diff
@@ -0,0 +1,99 @@
+            case
+            when diff > 0
+              items.concat(["<ul>"] * diff)
+            when diff < 0
```
Disable in .rubocop.yml

mtsmfm commented on lib/daimon_markdown/plugins/toc_pages.rb:
```diff
@@ -0,0 +1,107 @@
+        page_separated_nodes = []
+        condition = 1.upto(6).map {|n| "./following::h#{n}" }.join("|")
+        html.search("//comment()[contains(.,\"nextpage\")]").map do |comment_element|
+          page_separated_nodes << comment_element.search(condition).first
```
```ruby
page_separated_nodes = html.search("//comment()[contains(.,\"nextpage\")]").map do |comment_element|
  comment_element.search(condition).first
end
```

okkez replied:
Ah! Good catch!

okkez committed:
Use Array#map properly

okkez committed:
Remove redundant assignment

mtsmfm commented on lib/daimon_markdown/plugins/toc_pages.rb:
```diff
@@ -0,0 +1,107 @@
+            end
+          headers[id] += 1
+          header_content = header_node.children.first
+          # TODO: Arrange indent level
```
TODO?

okkez replied:
This is for generated html indentation.
It is not important.

okkez committed:
Use `#at` instead of `#search` with first

`#at` returns first element which is matched condition.

okkez committed:
Add Header, UnorderedList, ListItem to simplify code

okkez committed:
Fix for rubocop suggestions

okkez committed:
Merge branch 'master' into toc-pages

okkez committed:
Generate nested ul-li list properly

ul must not be child of ul directry.

okkez committed:
Remove empty else clause

okkez committed:
Merge branch 'master' into toc-pages

mtsmfm commented:
I got 500 error with following body:

```
{{toc_pages}}

# Hello
hi !

<!--nextpage -->

## h2

a

### h3

b
```

okkez commented:
```
ActionView::Template::Error (undefined method `[]' for nil:NilClass):
  24:           = render 'posts/credits', credits: @post.credits.with_ordered at=info method=GET path="/editor/posts/100177/preview" host=daimon-news-stg-pr-691.herokuapp.com request_id=66c161f0-e9cf-4b5a-9d96-4c6343c4964d fwd="121.102.114.25" dyno=web.1 connect=2ms service=133ms status=500 bytes=1776
  25:       %p.thumbnail= image_tag @post.thumbnail_url
  26:     .post__body
  27:       = render_markdown(@pages.map(&:body).join, @markdown_context)
  28: 
  29:     - if @pages.total_pages > 1
  30:       %nav.pagination
lib/daimon_markdown/plugins/toc_pages.rb:57:in `block in call'
lib/daimon_markdown/plugins/toc_pages.rb:56:in `call'
lib/daimon_markdown/plugins/toc_pages.rb:56:in `zip'
app/helpers/markdown_helper.rb:4:in `render_markdown'
app/views/posts/show.html.haml:27:in `block in _app_views_posts_show_html_haml___3001049241409430540_70268641750820'
app/views/posts/show.html.haml:13:in `_app_views_posts_show_html_haml___3001049241409430540_70268641750820'
config/environments/production.rb:100:in `call'
app/controllers/editor/posts_controller.rb:66:in `preview'
app/middlewares/redirector.rb:25:in `response'
config/application.rb:44:in `call'
app/middlewares/redirector.rb:7:in `call'
```

okkez committed:
Check number of headers in the page

This change needs daimon_markdown 0.7.0 or later

okkez committed:
Add missing space around `=` operator
